Compatibility-2.0.md

```diff
@@ -26,3 +26,304 @@ Embedded Mode also might use a decorated Cache using a Transcoder. For example t
+* HotRod: there is no such capability yet, but it's something that will eventually be added.
+
+
+How does the Mimetype relate to the encoder general mechanism?
+
+
+## Proposal
+
+The proposal is to use a pair (Encoder, MimeType) to describe cache storage. Examples: 
+
+* "```application/json``` with no encoding" means IdentityEncoder associated of a content that represents a json (```byte[]```, ```Gson```, ```Map<String, Map<?,?>>```, etc)
+* "```text/plain; charset=utf-8``` with marshalling" means content produced by the MarshallerEncoder that represents text, so it's a ```byte[]```
+* "```application/octet-stream``` with KryoEncoder" means ```byte[]``` produced by a KryoMarshaller.
+
+This means that different MimeTypes can be associated with the same Encoder, e.g. the IdentityEncoder can be associated with any MimeType meaning whatever is sent
+to cache will be written or read without transformations.
+
+The purpose of the MimeType is twofold:
+
+* Provide a metadata about the content of the cache, independently of the encoding;
+* By describing the format, enable the conversion between different encodings, a.k.a. Transcoding.
+
+
+### Configuration changes
+
+A cache will have a data-type section that looks like the following:
+
+```xml
+<local-cache>
+  <data-type>
+     <key mime-type="application/xml" encoder="org.infinispan.commons.dataconversion.IdentityEncoder"/>
+     <value mime-type="text/plain; charset=UTF-8" encoder="org.infinispan.dataconversion.MarshalledEncoder"/>
+  </data-type>
+</local-cache>
+```
+
+The MimeType will be either present in the config or inferred; if not provided, Infinispan will automatically pick the best 
+decorated cache depending on the situation.
+
+
+### API changes
+Apart from being able to use a different encoders on the fly, it will be possible to ask for a different MimeType using the method:
+
+```java
+// Both keys and types as "application/xml"
+cache.getAdvancedCache.withMimeType("application/xml").withEncoding(IdentityEncoder.class)
+
+// Different MimeTypes for keys and values:
+cache.getAdvancedCache.withMimeType("application/x-java-object", "application/json").withEncoding(IdentityEncoder.class)
+
+```
+
+Internally the cache will compare the passed in mimetype with the configured mimetype from the cache and will perform a transcode operation if they 
+differ.
+
+### Implementation
+
+The transcode operation will take a tuple (Object, MimeType) plus a destination MimeType and will return a converted Object that will be sent to the encoder.
+
+```java
+import javax.activation.MimeType;
+
+public interface Transcoder {
+
+   /**
+    * Transcode content between two different {@link MimeType}
+    *
+    * @param content         Content to transcode
+    * @param contentType     The {@link MimeType} of the content
+    * @param destinationType The target {@link MimeType} to convert to
+    * @param binary          if true, produces byte[] as output
```

Comment:
Couldn't you use the configured encoder to define what the returned type is? Wouldn't that give more flexibility that just a boolean that defines whether a byte[] is produced?